[OM] pthread barriers and spinlocks are unusable: types hidden behind __USE_XOPEN2K and unmodelled

### Summary

Programs using POSIX barriers or spinlocks cannot be verified at all: the types are declared behind `#ifdef __USE_XOPEN2K`, which ESBMC never defines, and there is no operational model for `pthread_barrier_wait` / `pthread_spin_lock` even if the types were visible.

Split out of #6475, where a barrier deadlock test could not be written.

### Reproducer

```c
/* Barrier of 3, but only 2 threads ever arrive: a real deadlock. */
#include <pthread.h>
pthread_barrier_t bar;

void *t(void *p) { pthread_barrier_wait(&bar); return 0; }

int main(void)
{
  pthread_t a, b;
  pthread_barrier_init(&bar, 0, 3);
  pthread_create(&a, 0, t, 0);
  pthread_create(&b, 0, t, 0);
  pthread_join(a, 0);
  pthread_join(b, 0);
  return 0;
}
```

```
$ esbmc v2_barrier.c --deadlock-check --unwind 4
v2_barrier.c:3:1: error: unknown type name 'pthread_barrier_t'; did you mean 'pthread_attr_t'?
ERROR: PARSING ERROR
```

Adding `-D_GNU_SOURCE` does not help.

### Root cause

`src/c2goto/headers/pthread.h:106-122`:

```c
#ifdef __USE_XOPEN2K
/* POSIX spinlock data type.  */
typedef volatile int pthread_spinlock_t;

typedef union { long int __align; } pthread_barrier_t;
typedef union { int __align; }      pthread_barrierattr_t;
#endif
```

Nothing in ESBMC's header set defines `__USE_XOPEN2K` (unlike glibc, which defines it from `features.h` based on `_POSIX_C_SOURCE >= 200112L`). The declarations at lines 721+ (`pthread_barrier_init`, `pthread_barrier_destroy`, `pthread_barrier_wait`, and the `pthread_spin_*` family) are inside the same guard.

Independently, `grep -n barrier src/c2goto/library/*.c` returns nothing: there is no operational model for any of these functions, so even with the types exposed they would be unmodelled external calls with no synchronisation semantics — barrier deadlocks would be missed and the barrier's ordering guarantee would not constrain interleavings, which also affects `--data-races-check`.

### Suggested fix

Expose the types (define `__USE_XOPEN2K` in the header prelude, or drop the guard as was already done for the rwlock block just above it), then add a barrier model to `pthread_lib.c`: a count of arrived threads and a required count, with the blocked branch doing the same `__ESBMC_blocked_threads_count` accounting and `blocked != running` assertion as `pthread_mutex_lock_check`. See #6474 for the caveat that this counter is itself currently unsound.

### Environment

ESBMC 8.4.0, master @ 1a6dbf32d5, macOS aarch64.

--- a/regression/esbmc-unix/github_6478-spinlock/main.c
+++ b/regression/esbmc-unix/github_6478-spinlock/main.c
@@ -1,0 +1,26 @@
+#include <pthread.h>
+#include <assert.h>
+pthread_spinlock_t sl;
+int shared = 0;
+
+void *t(void *p)
+{
+  pthread_spin_lock(&sl);
+  shared++;
+  pthread_spin_unlock(&sl);
+  return 0;
+}
+
+int main(void)
+{
+  pthread_t a;
+  pthread_spin_init(&sl, 0);
+  pthread_create(&a, 0, t, 0);
+  pthread_spin_lock(&sl);
+  shared++;
+  pthread_spin_unlock(&sl);
+  pthread_join(a, 0);
+  assert(shared == 2);
+  pthread_spin_destroy(&sl);
+  return 0;
+}

--- a/regression/esbmc-unix/github_6478-spinlock/test.desc
+++ b/regression/esbmc-unix/github_6478-spinlock/test.desc
@@ -1,0 +1,4 @@
+CORE
+main.c
+--unwind 4 --deadlock-check
+^VERIFICATION SUCCESSFUL$

--- a/regression/esbmc-unix/github_6478-spinlock_fail/main.c
+++ b/regression/esbmc-unix/github_6478-spinlock_fail/main.c
@@ -1,0 +1,10 @@
+/* Self-deadlock: a spinlock is not recursive. */
+#include <pthread.h>
+pthread_spinlock_t sl;
+int main(void)
+{
+  pthread_spin_init(&sl, 0);
+  pthread_spin_lock(&sl);
+  pthread_spin_lock(&sl);
+  return 0;
+}

--- a/regression/esbmc-unix/github_6478-spinlock_fail/test.desc
+++ b/regression/esbmc-unix/github_6478-spinlock_fail/test.desc
@@ -1,0 +1,5 @@
+CORE
+main.c
+--unwind 4 --deadlock-check
+^VERIFICATION FAILED$
+Deadlocked state in pthread_spin_lock

--- a/regression/esbmc-unix/github_6478/main.c
+++ b/regression/esbmc-unix/github_6478/main.c
@@ -1,0 +1,23 @@
+/* The barrier orders the write in the worker before the read in main. */
+#include <pthread.h>
+#include <assert.h>
+pthread_barrier_t bar;
+int shared = 0;
+
+void *t(void *p)
+{
+  shared = 42;
+  pthread_barrier_wait(&bar);
+  return 0;
+}
+
+int main(void)
+{
+  pthread_t a;
+  pthread_barrier_init(&bar, 0, 2);
+  pthread_create(&a, 0, t, 0);
+  pthread_barrier_wait(&bar);
+  assert(shared == 42);
+  pthread_join(a, 0);
+  return 0;
+}

--- a/regression/esbmc-unix/github_6478/test.desc
+++ b/regression/esbmc-unix/github_6478/test.desc
@@ -1,0 +1,4 @@
+CORE
+main.c
+--unwind 4 --deadlock-check
+^VERIFICATION SUCCESSFUL$

--- a/regression/esbmc-unix/github_6478_fail/main.c
+++ b/regression/esbmc-unix/github_6478_fail/main.c
@@ -1,0 +1,16 @@
+/* Barrier of 3, but only 2 threads ever arrive: a real deadlock. */
+#include <pthread.h>
+pthread_barrier_t bar;
+
+void *t(void *p) { pthread_barrier_wait(&bar); return 0; }
+
+int main(void)
+{
+  pthread_t a, b;
+  pthread_barrier_init(&bar, 0, 3);
+  pthread_create(&a, 0, t, 0);
+  pthread_create(&b, 0, t, 0);
+  pthread_join(a, 0);
+  pthread_join(b, 0);
+  return 0;
+}

--- a/regression/esbmc-unix/github_6478_fail/test.desc
+++ b/regression/esbmc-unix/github_6478_fail/test.desc
@@ -1,0 +1,5 @@
+CORE
+main.c
+--unwind 4 --deadlock-check
+^VERIFICATION FAILED$
+Deadlocked state in

--- a/src/c2goto/headers/pthread.h
+++ b/src/c2goto/headers/pthread.h
@@ -106,23 +106,33 @@ typedef union
 //#endif
 
 
-#ifdef __USE_XOPEN2K
-/* POSIX spinlock data type.  */
-typedef volatile int pthread_spinlock_t;
+/* Always exposed: ESBMC never defines __USE_XOPEN2K, so guarding these made
+   barriers and spinlocks unusable outright (#6478). */
+/* POSIX spinlock data type. Both fields are ESBMC bookkeeping: __lock holds
+   the state and __waiters the blocked-thread count, as for pthread_mutex_t. */
+typedef struct
+{
+  int __lock;
+  unsigned int __waiters;
+} pthread_spinlock_t;
 
 
 /* POSIX barriers data type.  The structure of the type is
    deliberately not exposed.  */
-typedef union
+typedef struct
 {
-  long int __align;
+  unsigned int __count;
+  unsigned int __arrived;
+  /* Bumped every time the barrier opens, so a waiter can tell that the
+     round it arrived in has completed. */
+  unsigned int __generation;
+  unsigned int __waiters;
 } pthread_barrier_t;
 
 typedef union
 {
   int __align;
 } pthread_barrierattr_t;
-#endif
 
 struct sched_param
 {
@@ -256,12 +266,10 @@ enum
 #define PTHREAD_ONCE_INIT 0
 
 
-#ifdef __USE_XOPEN2K
 /* Value returned by 'pthread_barrier_wait' for one of the threads after
    the required number of threads have called this function.
    -1 is distinct from 0 and all errno constants */
-# define PTHREAD_BARRIER_SERIAL_THREAD -1
-#endif
+#define PTHREAD_BARRIER_SERIAL_THREAD -1
 
 
 /* Create a new thread, starting with execution of START-ROUTINE
@@ -701,7 +709,7 @@ extern int pthread_condattr_setclock (pthread_condattr_t *__attr,
 #endif
 
 
-#ifdef __USE_XOPEN2K
+/* Always exposed; see the type definitions above. */
 /* Functions to handle spinlocks.  */
 
 /* Initialize the spinlock LOCK.  If PSHARED is nonzero the spinlock can
@@ -750,7 +758,6 @@ extern int pthread_barrierattr_getpshared (__const pthread_barrierattr_t *
 /* Set the process-shared flag of the barrier attribute ATTR.  */
 extern int pthread_barrierattr_setpshared (pthread_barrierattr_t *__attr,
 					   int __pshared);
-#endif
 
 
 /* Functions for handling thread-specific data.  */

--- a/src/c2goto/library/pthread_lib.c
+++ b/src/c2goto/library/pthread_lib.c
@@ -29,6 +29,12 @@ void __ESBMC_set_thread_internal_data(
 #define __ESBMC_rwlock_readers(a) ((a)->__readers)
 #define __ESBMC_rwlock_writer(a) ((a)->__writer)
 #define __ESBMC_rwlock_waiters(a) ((a)->__waiters)
+#define __ESBMC_barrier_count(a) ((a)->__count)
+#define __ESBMC_barrier_arrived(a) ((a)->__arrived)
+#define __ESBMC_barrier_generation(a) ((a)->__generation)
+#define __ESBMC_barrier_waiters(a) ((a)->__waiters)
+#define __ESBMC_spin_lock_field(a) ((a)->__lock)
+#define __ESBMC_spin_waiters(a) ((a)->__waiters)
 
 /* Global tracking data. Should all initialize to 0 / false */
 __attribute__((annotate("__ESBMC_inf_size")))
@@ -721,6 +727,175 @@ __ESBMC_HIDE:;
 
   __ESBMC_assume(acquired);
 
+  return 0;
+}
+
+/************************ barrier mainpulation routines ***********************/
+
+int pthread_barrierattr_init(pthread_barrierattr_t *attr)
+{
+  return 0;
+}
+
+int pthread_barrierattr_destroy(pthread_barrierattr_t *attr)
+{
+  return 0;
+}
+
+int pthread_barrierattr_getpshared(
+  const pthread_barrierattr_t *attr,
+  int *pshared)
+{
+__ESBMC_HIDE:;
+  *pshared = PTHREAD_PROCESS_PRIVATE;
+  return 0;
+}
+
+int pthread_barrierattr_setpshared(pthread_barrierattr_t *attr, int pshared)
+{
+  return 0;
+}
+
+int pthread_barrier_init(
+  pthread_barrier_t *barrier,
+  const pthread_barrierattr_t *attr,
+  unsigned int count)
+{
+__ESBMC_HIDE:;
+  if (count == 0)
+    return EINVAL;
+
+  __ESBMC_atomic_begin();
+  __ESBMC_barrier_count(barrier) = count;
+  __ESBMC_barrier_arrived(barrier) = 0;
+  __ESBMC_barrier_generation(barrier) = 0;
+  __ESBMC_release_blocked_threads(__ESBMC_barrier_waiters(barrier));
+  __ESBMC_barrier_waiters(barrier) = 0;
+  __ESBMC_atomic_end();
+  return 0;
+}
+
+int pthread_barrier_destroy(pthread_barrier_t *barrier)
+{
+__ESBMC_HIDE:;
+  __ESBMC_atomic_begin();
+  __ESBMC_assert(
+    __ESBMC_barrier_arrived(barrier) == 0,
+    "attempt to destroy a barrier with waiting threads");
+  __ESBMC_barrier_count(barrier) = 0;
+  __ESBMC_atomic_end();
+  return 0;
+}
+
+int pthread_barrier_wait(pthread_barrier_t *barrier)
+{
+__ESBMC_HIDE:;
+  __ESBMC_atomic_begin();
+
+  unsigned int generation = __ESBMC_barrier_generation(barrier);
+  int res = 0;
+
+  __ESBMC_barrier_arrived(barrier)++;
+
+  if (__ESBMC_barrier_arrived(barrier) == __ESBMC_barrier_count(barrier))
+  {
+    // Last to arrive: open the barrier and let this round's waiters through.
+    __ESBMC_barrier_arrived(barrier) = 0;
+    __ESBMC_barrier_generation(barrier)++;
+    __ESBMC_release_blocked_threads(__ESBMC_barrier_waiters(barrier));
+    __ESBMC_barrier_waiters(barrier) = 0;
+    res = PTHREAD_BARRIER_SERIAL_THREAD;
+  }
+  else
+  {
+    __ESBMC_barrier_waiters(barrier)++;
+    __ESBMC_blocked_threads_count++;
+    __ESBMC_assert(
+      __ESBMC_blocked_threads_count != __ESBMC_num_threads_running,
+      "Deadlocked state in pthread_barrier_wait");
+  }
+
+  __ESBMC_atomic_end();
+
+  // Only schedules in which this thread's round has since completed continue.
+  __ESBMC_assume(__ESBMC_barrier_generation(barrier) != generation);
+
+  return res;
+}
+
+/*********************** spinlock mainpulation routines ***********************/
+
+int pthread_spin_init(pthread_spinlock_t *lock, int pshared)
+{
+__ESBMC_HIDE:;
+  __ESBMC_atomic_begin();
+  __ESBMC_spin_lock_field(lock) = 0;
+  __ESBMC_release_blocked_threads(__ESBMC_spin_waiters(lock));
+  __ESBMC_spin_waiters(lock) = 0;
+  __ESBMC_atomic_end();
+  return 0;
+}
+
+int pthread_spin_destroy(pthread_spinlock_t *lock)
+{
+  return 0;
+}
+
+int pthread_spin_lock(pthread_spinlock_t *lock)
+{
+__ESBMC_HIDE:;
+  __ESBMC_atomic_begin();
+
+  _Bool acquired = (__ESBMC_spin_lock_field(lock) == 0);
+
+  if (acquired)
+  {
+    __ESBMC_spin_lock_field(lock) = 1;
+  }
+  else
+  {
+    // A spinning thread never yields, so every thread spinning at once is a
+    // deadlock exactly as it is for a blocking mutex.
+    __ESBMC_spin_waiters(lock)++;
+    __ESBMC_blocked_threads_count++;
+    __ESBMC_assert(
+      __ESBMC_blocked_threads_count != __ESBMC_num_threads_running,
+      "Deadlocked state in pthread_spin_lock");
+  }
+
+  __ESBMC_atomic_end();
+
+  __ESBMC_assume(acquired);
+
+  return 0;
+}
+
+int pthread_spin_trylock(pthread_spinlock_t *lock)
+{
+__ESBMC_HIDE:;
+  __ESBMC_atomic_begin();
+
+  int res = EBUSY;
+  if (__ESBMC_spin_lock_field(lock) == 0)
+  {
+    __ESBMC_spin_lock_field(lock) = 1;
+    res = 0;
+  }
+
+  __ESBMC_atomic_end();
+  return res;
+}
+
+int pthread_spin_unlock(pthread_spinlock_t *lock)
+{
+__ESBMC_HIDE:;
+  __ESBMC_atomic_begin();
+  __ESBMC_assert(
+    __ESBMC_spin_lock_field(lock), "must hold spinlock upon unlock");
+  __ESBMC_spin_lock_field(lock) = 0;
+  __ESBMC_release_blocked_threads(__ESBMC_spin_waiters(lock));
+  __ESBMC_spin_waiters(lock) = 0;
+  __ESBMC_atomic_end();
   return 0;
 }
 

